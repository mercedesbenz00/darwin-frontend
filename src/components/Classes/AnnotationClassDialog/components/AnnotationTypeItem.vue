<template>
  <div
    v-if="annotationType"
    v-tooltip="disabled ? {
      content: disabledTooltip,
      placement: 'left',
      offset: 10
    } : undefined"
    class="annotation-type-item"
    :class="{
      'annotation-type-item--disabled': disabled,
      'annotation-type-item--selected': selected
    }"
    @click="onClick"
  >
    <type-icon
      class="annotation-type-item__icon"
      :type="typeName"
      :color="iconColor"
    />
    <span class="annotation-type-item__label">{{ typeNameFormatted }}</span>
    <!--
      The v-if here is because there is no tooltip defined for the "link" type,
      so any team which has links enabled will get errors otherwise.
      https://linear.app/v7labs/issue/V7-985/add-design-for-link-tooltip-in-annotation-class-dialog
    -->
    <annotation-type-item-info
      v-if="hasInfo"
      class="annotation-type-item__info"
      :data="annotationType"
    />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import TypeIcon from '@/components/Common/AnnotationType/TypeIcon.vue'
import {
  AnnotationTypeName,
  AnnotationTypePayload,
  FeaturePayload,
  RootState
} from '@/store/types'
import { annotationTypeRichData, AnnotationTypeRichPayload, formatTypeName } from '@/utils'

import AnnotationTypeItemInfo from './AnnotationTypeItemInfo/AnnotationTypeItemInfo.vue'

@Component({
  name: 'annotation-type-item',
  components: { AnnotationTypeItemInfo, TypeIcon }
})
export default class AnnotationTypeItem extends Vue {
  @Prop({ required: true })
  typeName!: AnnotationTypeName

  get typeNameFormatted (): string {
    return formatTypeName(this.typeName)
  }

  @Prop({ type: Boolean, default: false })
  disabled!: boolean

  @Prop({ default: null })
  disabledTooltip!: string | null

  @Prop({ type: Boolean, default: false })
  selected!: boolean

  get iconColor (): string {
    return this.$theme.getColor('colorFeatherLight')
  }

  get hasInfo (): boolean {
    const typesWithNoInfo: AnnotationTypeName[] = ['graph', 'link', 'string', 'table']
    return !typesWithNoInfo.includes(this.typeName)
  }

  onClick (): void {
    if (this.disabled) { return }
    this.$emit('click', this.typeName)
  }

  @State((state: RootState) => state.features.list)
  features!: FeaturePayload[] | null

  @State((state: RootState) => state.aclass.types)
  annotationTypes!: AnnotationTypePayload[]

  /**
   * The following object maps annotation type names to the features
   * name that enables them.
   */
  featureEnabledTypes = {
    graph: 'DOCUMENTS',
    link: 'LINK_TOOL',
    string: 'DOCUMENTS',
    mask: 'RASTERS',
    table: 'DOCUMENTS'
  }

  /**
   * Computed property which is used to figure out with rich annotation types should be included
   * when some annotation types are guarded behind a feature flag.
   *
   * The return type is an object with the hidden annotation type name as key,
   * and the rich annotation type payload to be shown as value.
   */
  get annotationType (): AnnotationTypeRichPayload | null {
    const { annotationTypes, features, typeName } = this

    for (const [featureEnabledTypeName, featureName] of Object.entries(this.featureEnabledTypes)) {
      if (typeName !== featureEnabledTypeName) { continue }

      const feature = features?.find(f => f.name === featureName)
      if (!feature?.enabled) { return null }
    }

    const type = annotationTypes.find(t => t.name === typeName)
    const richType = annotationTypeRichData.find(r => r.name === typeName)

    if (!type || !richType) { return null }

    const richSubs = type.subs.map((sub) => {
      const richSub = annotationTypeRichData.find(r => r.name === typeName) || {}
      if (sub && richSub) {
        return { ...sub, ...richSub }
      }
      return null
    }).filter((sub): sub is AnnotationTypeRichPayload => !!sub)

    if (type && richType) {
      return {
        ...type,
        ...richType,
        richSubs
      }
    }

    return null
  }
}
</script>

<style lang="scss" scoped>
.annotation-type-item {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: 13px;
  padding: 2px 3px;
  border: 1px solid $colorSecondaryLight3;
  transition: background-color .2s;
  cursor: pointer;
  @include noSelect;

  &:hover:not(.annotation-type-item--disabled),
  &:active:not(.annotation-type-item--disabled) {
    border: 1px solid $colorSecondaryLight2;
    background-color: $colorSecondaryLight2;
    transition: none;
  }
}

.annotation-type-item--selected {
  border: 1px solid $colorSecondaryLight;
  background-color: $colorSecondaryLight1;

  &:hover:not(.annotation-type-item--disabled),
  &:active:not(.annotation-type-item--disabled) {
    border: 1px solid $colorSecondaryLight;
    background-color: $colorSecondaryLight1;
  }
}

.annotation-type-item--disabled {
  .annotation-type-item__icon {
    opacity: .3;
  }
  .annotation-type-item__label {
    opacity: .3;
  }
}

.annotation-type-item__icon {
  @include row--center;
  width: 23px;
  height: 23px;
  margin: 0 8px 0 3px;
  object-fit: contain;
}

.annotation-type-item__label {
  flex: 1;
  @include typography(md, default, normal);
  text-align: left;
  color: $colorSecondaryDark;
  white-space: nowrap;
}
</style>
