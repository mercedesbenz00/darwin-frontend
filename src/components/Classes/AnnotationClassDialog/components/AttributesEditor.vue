<template>
  <annotation-class-section title="Attributes">
    <div class="attributes-container">
      <vue-tags-input
        ref="attributeTagInput"
        v-model="attributeTag"
        avoid-adding-duplicates
        allow-edit-tags
        :add-on-blur="false"
        :delete-on-backspace="false"
        :disabled="!editing || loading"
        :placeholder="attributeInputPlaceholder"
        :tags="attributeTags"
        @before-adding-tag="attributeOnNewTag"
        @before-editing-tag="attributeOnEditTag"
        @before-saving-tag="attributeOnSaveTag"
        @tags-changed="attributeTagsChanged"
      >
        <div
          slot="tag-actions"
          slot-scope="props"
        >
          <v-popover
            trigger="manual"
            :open="attributeMenu[props.tag.id]"
            placement="right"
            popover-class="attributes-menu"
            @hide="onHideAttributeMoreMenu(props)"
          >
            <div
              class="ti-icon-more"
              @click="showAttributesMenu(props)"
            />
            <template slot="popover">
              <div
                class="attributes-menu__item"
                @click="maybeRecolorAttribute(props)"
              >
                Recolor
              </div>
              <div
                class="attributes-menu__item"
                @click="maybeRenameAttribute(props)"
              >
                Rename
              </div>
              <div
                class="attributes-menu__item"
                @click="maybeDeleteAttribute(props)"
              >
                Delete
              </div>
            </template>
          </v-popover>
          <v-popover
            trigger="manual"
            :open="attributeColorMenu[props.tag.id]"
            placement="right"
            popover-class="tooltip--white popover--colors popover--attribute-colors"
            @hide="onHideAttributeColorMenu(props)"
          >
            <div class="ti-icon-color" />
            <template slot="popover">
              <color-picker
                :colors="annotationClassColors"
                @select="onSelectAttributeColor(props, $event)"
              />
            </template>
          </v-popover>
        </div>
      </vue-tags-input>
    </div>

    <delete-confirmation-dialog
      ref="deleteAttributeDialog"
      name="delete-attribute-modal"
      button-text="REMOVE ATTRIBUTE"
      :title="deleteAttributeDialogTitle"
      :detail="deleteAttributeDialogDetails"
      @confirmed="deleteAttribute"
    />
  </annotation-class-section>
</template>

<script lang="ts">
import VueTagsInput, { VueTagsInputProp } from '@johmun/vue-tags-input'
import { Component, Prop, Vue, Watch } from 'vue-property-decorator'
import { Getter } from 'vuex-class'

import ColorPicker from '@/components/Classes/ColorPicker/ColorPicker.vue'
import { DeleteConfirmationDialog } from '@/components/Common/DeleteConfirmationDialog/V1'
import { AttributePayload, InputTag } from '@/store/types'
import {
  errorsByCode,
  getAttributeColor
} from '@/utils'
import { annotationClassColors as v2AnnotationClassColors } from '@/utilsV2'

import AnnotationClassSection from './Common/AnnotationClassSection.vue'

@Component({
  name: 'attributes-editor',
  components: { AnnotationClassSection, ColorPicker, DeleteConfirmationDialog, VueTagsInput }
})
export default class AttributesEditor extends Vue {
  @Prop()
  annotationClassId?: number

  @Prop({ type: Boolean, default: false })
  editing!: boolean

  attributeTags: InputTag[] = []
  attributeTag = ''
  attributeMenu: Record<string, boolean> = {}
  attributeColorMenu: Record<string, boolean> = {}
  attributeTagToDelete: InputTag | null = null

  loading: boolean = false

  $refs!: {
    attributeTagInput: InstanceType<typeof VueTagsInput>
    deleteAttributeDialog: DeleteConfirmationDialog
  }

  annotationClassColors = v2AnnotationClassColors

  get attributeInputPlaceholder (): string {
    return this.editing
      ? 'Type in an attribute'
      : 'Save the annotation class to edit attributes'
  }

  get deleteAttributeDialogTitle (): string {
    if (!this.attributeTagToDelete) { return '' }
    return `Please confirm you wish to remove "${this.attributeTagToDelete.text}" attribute.`
  }

  get deleteAttributeDialogDetails (): string {
    if (!this.attributeTagToDelete) { return '' }
    const { text } = this.attributeTagToDelete
    return `This will remove the "${text}" attribute from the current annotation class.`
  }

  @Getter('attributesByAnnotationClassId', { namespace: 'aclass' })
  attributesByAnnotationClassId!: {
    [key: string]: AttributePayload[]
  }

  get assignedAttributeTags (): AttributePayload[] {
    if (!this.annotationClassId) { return [] }
    return this.attributesByAnnotationClassId[this.annotationClassId!]
  }

  get canCreateAttributes (): boolean {
    return this.$can('update_annotation_class')
  }

  /**
   * Enforce focus to the vue-tags-input input element
   */
  setFocusToAttributeInput (): void {
    const tagInputEl = this.$el.querySelector('.ti-new-tag-input') as (HTMLElement | null)
    if (!tagInputEl) { return }
    this.$nextTick(() => tagInputEl.focus())
  }

  async attributeOnNewTag (object: {tag: InputTag, addTag: () => void}): Promise<void> {
    if (!this.annotationClassId) { return }
    if (!this.canCreateAttributes) {
      return this.$store.dispatch(
        'toast/warning',
        { content: errorsByCode.ANNOTATION_CREATE_ATTRIBUTE_NOT_AUTHORIZED }
      )
    }
    if (object.tag.tiClasses && object.tag.tiClasses!.includes('ti-duplicate')) {
      this.attributeTag = ''
      return
    }

    const params = {
      classId: this.annotationClassId,
      name: object.tag.text,
      color: getAttributeColor(object.tag.text)
    }
    this.loading = true
    const {
      data,
      error
    } = await this.$store.dispatch('aclass/createOrFetchAnnotationAttribute', params)

    this.loading = false

    if (error) {
      this.attributeTag = ''
      return
    }

    // store the id instead of the text in the attribute viewer
    const tagInfo = this.tagFromAttribute(data)
    object.tag.id = tagInfo.id
    object.tag.text = tagInfo.text
    object.tag.style = tagInfo.style
    this.attributeTag = ''
    object.addTag()

    this.setFocusToAttributeInput()
  }

  attributeOnEditTag (object: {tag: InputTag, editTag: () => void}): void {
    object.editTag()
  }

  async attributeOnSaveTag (object: {tag: InputTag, saveTag: () => void}): Promise<void> {
    if (!this.annotationClassId) { return }
    if (!this.canCreateAttributes) {
      return this.$store.dispatch(
        'toast/warning',
        { content: errorsByCode.ANNOTATION_UPDATE_ATTRIBUTE_NOT_AUTHORIZED }
      )
    }

    const params = {
      classId: this.annotationClassId,
      id: object.tag.id,
      name: object.tag.text,
      color: getAttributeColor(object.tag.text)
    }
    this.loading = true
    const { data, error } = await this.$store.dispatch('aclass/updateAnnotationAttribute', params)
    this.loading = false
    if (error) { return }

    const tagInfo = this.tagFromAttribute(data)
    object.tag.style = tagInfo.style
    object.saveTag()
  }

  attributeTagsChanged (newTags: InputTag[]): void {
    if (!this.annotationClassId) { return }
    this.attributeTags = newTags
  }

  tagFromAttribute (attribute: AttributePayload): InputTag {
    return {
      id: attribute.id,
      text: attribute.name,
      style: `background: ${attribute.color}`
    }
  }

  @Watch('assignedAttributeTags')
  onAssignedAttributeTagsChange (): void {
    if (!this.assignedAttributeTags) {
      this.attributeTags = []
      return
    }
    this.attributeTags = this.assignedAttributeTags.map(this.tagFromAttribute)
  }

  showAttributesMenu (props: VueTagsInputProp): void {
    this.$set(this.attributeMenu, props.tag.id, true)
  }

  onHideAttributeMoreMenu (props: VueTagsInputProp): void {
    this.$set(this.attributeMenu, props.tag.id, false)
  }

  maybeRecolorAttribute (props: VueTagsInputProp): void {
    this.$set(this.attributeColorMenu, props.tag.id, true)
    this.onHideAttributeMoreMenu(props)
  }

  onHideAttributeColorMenu (props: VueTagsInputProp): void {
    this.$set(this.attributeColorMenu, props.tag.id, false)
  }

  async onSelectAttributeColor (props: VueTagsInputProp, color: string): Promise<void> {
    this.onHideAttributeColorMenu(props)
    const params = {
      classId: this.annotationClassId,
      id: props.tag.id,
      color
    }
    await this.$store.dispatch('aclass/updateAnnotationAttribute', params)
  }

  maybeRenameAttribute (props: VueTagsInputProp): void {
    this.$set(this.attributeMenu, props.tag.id, false)
    this.$refs.attributeTagInput.editTag(props.index)
  }

  maybeDeleteAttribute (props: VueTagsInputProp): void {
    this.$set(this.attributeMenu, props.tag.id!, false)

    this.attributeTagToDelete = props.tag
    this.$refs.deleteAttributeDialog.show()
  }

  async deleteAttribute (): Promise<void> {
    this.$refs.deleteAttributeDialog.close()
    this.$store.dispatch('ui/putBackSidebar')

    if (!this.attributeTagToDelete) { return }
    const params = {
      classId: this.annotationClassId,
      id: this.attributeTagToDelete.id!
    }
    await this.$store.dispatch('aclass/deleteAnnotationAttribute', params)
  }
}
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
@import "./assets/color-popover.scss";

.popover--attribute-colors {
  top: -9px !important;
}

.attributes-container {
  .vue-tags-input {
    width: 100%;
    max-width: unset !important;
    max-height: 200px;
    overflow: auto;

    border-radius: $border-radius-default;
    @include typography(md, default);
    background: $colorLineGrey !important;
    color: $colorSecondaryDark;

    input::-webkit-input-placeholder {
      color: $colorSecondaryLight;
    }
    input::-moz-placeholder {
      color: $colorSecondaryLight;
    }
    input:-moz-placeholder {
      color: $colorSecondaryLight;
    }
    input:-ms-input-placeholder {
      color: $colorSecondaryLight;
    }
  }

  .ti-input {
    @include typography(md, default);
    padding: 6px 4px !important;
    border: none !important;
    background: $colorLineGrey !important;
    box-shadow: inset 0px 2px 4px rgba(145, 169, 192, 0.3);
  }

  .ti-new-tag-input {
    @include typography(md, default);
    background: $colorLineGrey !important;
  }

  .ti-tag {
    @include typography(md, default);
    color: $colorSecondaryDark1 !important;
    border-radius: 5px !important;
    padding: 4px 5px 4px 7px !important;
  }

  .ti-actions {
    margin-left: 5px !important;

    .ti-icon-loading {
      background: url('/static/imgs/spinner.svg');
      background-size: 10px 10px;
      width: 10px;
      height: 10px;
      margin-right: -3px; // This is the margin to reduce the right space
      border: none;
      background-repeat: no-repeat;
    }

    i::before {
      color: rgba(0, 0, 0, 0.3);
    }

    i:hover {
      &::before {
        color: $colorSecondaryLight1;
      }
    }

    .ti-icon-more {
      width: 13px;
      height: 6px;
      padding: 0;
      background: url('/static/imgs/more-white-icon.svg');
      background-size: contain;
      background-repeat: no-repeat;
      cursor: pointer;
    }

    .ti-icon-color {
      width: 13px;
      height: 0;
      padding: 0;
    }

    .v-popover:last-child {
      height: 0;
    }
  }
}

.attributes-menu {
  @include noSelect;
  border-radius: 3px;
  overflow: hidden;
  box-shadow: 0px 20px 60px rgba(11, 36, 72, 0.2);

  .popover-inner {
    background: $colorWhite;
    border: none;
    padding: 5px 0;
  }

  .popover-arrow {
    display: none;
    background: $colorWhite;
  }

  .attributes-menu__item {
    color: $colorSecondaryDark1;
    padding: 5px 10px;
    cursor: pointer;
    @include typography(md);

    &:hover {
      background: $colorLineGrey;
    }
  }
}
</style>
