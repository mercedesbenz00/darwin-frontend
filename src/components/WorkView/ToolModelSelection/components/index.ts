export { default as ClassPill } from './ClassPill.vue'
export { default as ModelIcon } from './ModelIcon.vue'
