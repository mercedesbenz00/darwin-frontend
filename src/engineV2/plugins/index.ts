import attributesPlugin from './attributes'
import boundingBox from './boundingBox'
import boundingBox3d from './boundingBox3d'
import brush from './brush'
import clicker from './click'
import clipboardPlugin from './clipboard'
import commentatorPlugin from './commentator'
import directionalVector from './directionalVector'
import editPlugin from './edit'
import ellipse from './ellipse'
import fieldPlugin from './field'
import inferencePlugin from './inference'
import instanceIDPlugin from './instanceId'
import keypoint from './keypoint'
import link from './link'
import magnifierGlass from './magnifierGlass'
import mask from './mask'
import measuresPlugin from './measures'
import pixelBrush from './pixelBrush'
import polygon from './polygon'
import polyline from './polyline'
import saveAnnotations from './saveAnnotations'
import selectPlugin from './select'
import skeleton from './skeleton'
import tablePlugin from './table'
import tag from './tag'
import textPlugin from './text'
import v7Look from './v7'
import videoPlayerPlugin from './videoPlayer'
import windowLevelPlugin, { WindowLevelPlugin } from './windowLevel'
import zoomPlugin from './zoom'
import zoomCorrection from './zoomCorrection'
export {
  attributesPlugin,
  boundingBox,
  boundingBox3d,
  brush,
  clicker,
  clipboardPlugin,
  commentatorPlugin,
  directionalVector,
  editPlugin,
  ellipse,
  fieldPlugin,
  inferencePlugin,
  instanceIDPlugin,
  keypoint,
  link,
  magnifierGlass,
  mask,
  measuresPlugin,
  pixelBrush,
  polygon,
  polyline,
  saveAnnotations,
  selectPlugin,
  skeleton,
  tablePlugin,
  tag,
  textPlugin,
  v7Look,
  videoPlayerPlugin,
  windowLevelPlugin,
  WindowLevelPlugin,
  zoomPlugin,
  zoomCorrection
}
