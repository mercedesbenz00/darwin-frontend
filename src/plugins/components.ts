import Vue from 'vue'

import NegativeButton from '@/components/Common/Button/V1/NegativeButton.vue'
import PositiveButton from '@/components/Common/Button/V1/PositiveButton.vue'
import PrimaryButton from '@/components/Common/Button/V1/PrimaryButton.vue'
import SecondaryButton from '@/components/Common/Button/V1/SecondaryButton.vue'
import SecondaryLightButton from '@/components/Common/Button/V1/SecondaryLightButton.vue'

export const installCommonComponents = (_Vue: typeof Vue) => {
  _Vue.component('NegativeButton', NegativeButton)
  _Vue.component('PrimaryButton', PrimaryButton)
  _Vue.component('SecondaryButton', SecondaryButton)
  _Vue.component('SecondaryLightButton', SecondaryLightButton)
  _Vue.component('PositiveButton', PositiveButton)
}
