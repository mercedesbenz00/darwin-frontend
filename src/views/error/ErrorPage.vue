<template>
  <background>
    <div class="auth header">
      <logo />
    </div>
    <div class="center__content">
      <div class="error__modal">
        <div class="error__modal__title">
          Something went wrong
        </div>
        <div class="error__modal__text">
          You weren't supposed to end up here. Something must have gone wrong.
        </div>
        <div class="error__modal__text">
          You can contact support if you think it's a bug.
        </div>
        <div class="error__modal__button__center">
          <positive-button
            tag="a"
            href="mailto:support@v7labs.com"
            target="_blank"
          >
            Contact support
          </positive-button>
        </div>
      </div>
    </div>
  </background>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import Background from '@/components/Common/Background.vue'
import Logo from '@/components/Common/Logo.vue'

export default defineComponent({
  name: 'ErrorPage',
  components: { Background, Logo }
})
</script>

<style scoped lang="scss">
.error__modal {
  height: 380px;
  width: 520px;
  margin-top: 50px;
  margin-bottom: 40px;
  border-radius: 10px;
  background-color: $colorWhite;
  box-shadow: 0px 10px 60px rgba(11, 36, 72, 0.1), 0px 1px 2px rgba(11, 36, 72, 0.05);
}

.error__modal__title {
  margin: 50px 50px 25px 70px;
  @include typography(lg, default, bold);
  text-align: left;
  letter-spacing: 0.01em;
  color: $colorSecondaryDark1;
}

.error__modal__text {
  padding: 25px 50px 25px 70px;
}

.error__modal__button__center {
  @include row--center;
}
</style>
