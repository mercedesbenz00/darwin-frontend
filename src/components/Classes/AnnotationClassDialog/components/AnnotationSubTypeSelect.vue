<template>
  <annotation-class-section
    v-if="isRendering"
    title="Subtypes"
    optional
  >
    <div class="annotation-types">
      <annotation-type-item
        v-for="type in annotationSubtypes"
        :key="type"
        class="annotation-type-item"
        disabled-tooltip="Cannot remove existing subtypes"
        :type-name="type"
        :disabled="originalSubTypeSelections[type]"
        :selected="typeSelections[type]"
        @click="updateSubAnnotation(type)"
      />
    </div>
  </annotation-class-section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import Info from '@/components/Common/Info.vue'
import { AnnotationTypeName, AnnotationTypePayload, RootState } from '@/store/types'
import { formatTypeName } from '@/utils'

import AnnotationTypeItem from './AnnotationTypeItem.vue'
import AnnotationClassSection from './Common/AnnotationClassSection.vue'

const ORDERED_SUBS: AnnotationTypeName[] =
  ['attributes', 'text', 'directional_vector', 'instance_id']

@Component({
  name: 'annotation-sub-type-select',
  components: { AnnotationClassSection, AnnotationTypeItem, Info }
})
export default class AnnotationSubTypeSelect extends Vue {
  @Prop({ required: true })
  mainAnnotationType!: AnnotationTypeName | null

  /**
   * Once a subtype has been selected and persisted to backend,
   * it can no longer be removed from a class.
   *
   * This mapping is used to ensure this constraint.
   */
  @Prop({ required: true })
  originalSubTypeSelections!: Partial<Record<AnnotationTypeName, boolean>>

  @Prop({ required: true })
  typeSelections!: Partial<Record<AnnotationTypeName, boolean>>

  @State((state: RootState) => state.aclass.types)
  annotationTypes!: AnnotationTypePayload[]

  get annotationSubtypes (): AnnotationTypeName[] {
    const type =
      this.annotationTypes.find(t => t.name === this.mainAnnotationType && t.granularity === 'main')

    if (!type) {
      throw new Error('Invalid main type given as prop')
    }

    const typeSubs = type.subs.map(s => s.name)
    return ORDERED_SUBS.filter(s => typeSubs.includes(s))
  }

  get isRendering (): boolean {
    return this.annotationSubtypes.length > 0
  }

  formatTypeName = formatTypeName

  updateSubAnnotation (typeName: AnnotationTypeName) {
    this.$emit('update:typeSelections', {
      ...this.typeSelections,
      [typeName]: !this.typeSelections[typeName]
    })
  }
}
</script>

<style lang="scss" scoped>
.annotation-types {
  display: block;
  width: 100%;
}

.annotation-type-item {
  width: calc(50% - 10px);
  margin: 5px 0;

  &:nth-child(even) {
    margin-left: 10px;
  }

  &:nth-child(odd) {
    margin-right: 10px;
  }
}
</style>
