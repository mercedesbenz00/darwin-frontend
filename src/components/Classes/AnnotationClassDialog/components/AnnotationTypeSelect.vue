<template>
  <annotation-class-section
    ref="container"
    title="Annotation Type"
    :error="error"
  >
    <div class="annotation-types">
      <annotation-type-item
        v-for="type in mainAnnotationTypeNames"
        :key="type"
        class="annotation-type-item"
        disabled-tooltip="Cannot edit main annotation type"
        :type-name="type"
        :disabled="editing"
        :selected="selectedAnnotationType === type"
        @click="selectAnnotation(type)"
      />
    </div>
  </annotation-class-section>
</template>

<script lang="ts">
import { startCase } from 'lodash'
import { Component, Prop, Vue } from 'vue-property-decorator'
import { Getter } from 'vuex-class'

import Info from '@/components/Common/Info.vue'
import { AnnotationTypeName, AnnotationTypePayload } from '@/store/types'

import AnnotationTypeItem from './AnnotationTypeItem.vue'
import AnnotationClassSection from './Common/AnnotationClassSection.vue'

const ORDERED_TYPES: AnnotationTypeName[] = [
  'polygon',
  'bounding_box',
  'tag',
  'mask',
  'skeleton',
  'line',
  'keypoint',
  'ellipse',
  'cuboid',
  'link',
  'string',
  'graph',
  'table'
]

@Component({
  name: 'annotation-type-select',
  components: { AnnotationClassSection, AnnotationTypeItem, Info }
})
export default class AnnotationTypeSelect extends Vue {
  @Getter('mainAnnotationTypes', { namespace: 'aclass' })
  mainAnnotationTypes!: AnnotationTypePayload[]

  @Prop({ default: null, required: true })
  selectedAnnotationType!: AnnotationTypeName | null

  @Prop({ type: Boolean, default: false, required: false })
  editing!: boolean

  @Prop({ default: null, required: false })
  error!: string | null

  $refs!: {
    container: AnnotationClassSection
  }

  get mainAnnotationTypeNames (): AnnotationTypeName[] {
    const typeMains = this.mainAnnotationTypes.map(s => s.name)
    return ORDERED_TYPES.filter(s => typeMains.includes(s))
  }

  /**
   * Annotation type names are in `snake_case`.
   *
   * This function converts them to a format frendlier for display by
   * replacing underscores with spaces and capitalizing the first letter of
   * each word.
   *
   * This is called `startCase` in lodash
   */
  formatTypeName = startCase

  public scrollTo (): void {
    this.$refs.container.scrollTo()
  }

  selectAnnotation (typeName: AnnotationTypeName): void {
    if (this.error) {
      this.$emit('update:error', null)
    }
    this.$emit('update:selectedAnnotationType', typeName)
  }
}
</script>

<style lang="scss" scoped>
.annotation-types {
  display: block;
  width: 100%;
}

.annotation-type-item {
  width: calc(50% - 10px);
  margin: 5px 0;

  &:nth-child(even) {
    margin-left: 10px;
  }

  &:nth-child(odd) {
    margin-right: 10px;
  }
}
</style>
