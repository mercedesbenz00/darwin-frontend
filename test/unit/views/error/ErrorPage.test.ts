import { createLocalVue, shallowMount, Stubs } from '@vue/test-utils'

import { installCommonComponents } from '@/plugins/components'
import ErrorPage from '@/views/error/ErrorPage.vue'

const localVue = createLocalVue()
installCommonComponents(localVue)

it('matches snapshot', () => {
  const stubs: Stubs = { background: true, logo: true }
  const wrapper = shallowMount(ErrorPage, { localVue, stubs })
  expect(wrapper).toMatchSnapshot()
})
