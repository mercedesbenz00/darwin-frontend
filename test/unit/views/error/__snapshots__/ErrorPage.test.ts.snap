// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<background-stub>
  <div class="auth header">
    <logo-stub theme="dark" size="normal"></logo-stub>
  </div>
  <div class="center__content">
    <div class="error__modal">
      <div class="error__modal__title">
        Something went wrong
      </div>
      <div class="error__modal__text">
        You weren't supposed to end up here. Something must have gone wrong.
      </div>
      <div class="error__modal__text">
        You can contact support if you think it's a bug.
      </div>
      <div class="error__modal__button__center">
        <positive-button-stub tag="a" href="mailto:support@v7labs.com" target="_blank">
          Contact support
        </positive-button-stub>
      </div>
    </div>
  </div>
</background-stub>
`;
